sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -389,15 +384,17 @@
     this.doFnInvoker = DoFnInvokers.invokerFor(doFn);
     this.doFnInvoker.invokeSetup();
     this.stateBinder = new BeamFnStateBinder();
-    this.startBundleContext = new StartBundleContext();
-    this.processBundleContext = new ProcessBundleContext();
-    this.finishBundleContext = new FinishBundleContext();
     this.stateFinalizers = new ArrayList<>();
   }
 
   @Override
   public void startBundle() {
-    doFnInvoker.invokeStartBundle(startBundleContext);
+    doFnInvoker.invokeStartBundle(doFn.new StartBundleContext() {
```

Comment:
I like the clean-up but it would be best if we didn't create new context objects on each call to start/process/finish, please move dofn.new YYYBundleContext() { ... } into the FnApiDoFnRunner constructor.

Reply:
Done.